clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -145,31 +151,40 @@ public double compressionRate() {
     }
 
     /**
-     * Return the capacity of the buffer
+     * Return the capacity of the initial buffer, for writable records
+     * it may be different from the current buffer's capacity
      */
-    public int capacity() {
-        return this.capacity;
+    public int initialCapacity() {
+        return this.initCapacity;
     }
 
     /**
-     * Get the byte buffer that backs this records instance
+     * Get the byte buffer that backs this records instance for reading
      */
     public ByteBuffer buffer() {
+        if (writable)
+            throw new IllegalStateException("The memory records must not be writable any more before getting its underlying buffer");
+
         return buffer.duplicate();
     }
 
     /**
      * Return a flipped duplicate of the closed buffer to reading records
+     *
+     * NOTE: only used for tests
      */
     public ByteBuffer flip() {
         if (writable)
-            throw new IllegalStateException("The memory records need to be closed for write before rewinding for read");
+            throw new IllegalStateException("The memory records must not be writable any more before rewinding for read");
 
         return (ByteBuffer) buffer.flip();
     }
 
     @Override
     public Iterator<LogEntry> iterator() {
+        if (writable)
+            throw new IllegalStateException("The memory records must not be writable any more before iterating");
+
```

Comment:
Thinking about it a bit more. It will be weird to throw exceptions in toString() and it's hard to prevent toString() from being called before MemoryRecords is closed. We probably don't need to throw IllegalStateException here. If iterator() is called before close, the iterator will just return a snapshot of the records at the time the iterator is created.